src/irr/asset/CGLIWriter.cpp

```diff
@@ -195,18 +189,19 @@ namespace irr
 					const auto layerData = texture.data(gliLayer, gliFace, region->imageSubresource.mipLevel);
 					const auto sourceData = ptrBeginningOfRegion + (layer * layerSize);
 
-					for (size_t index = 0; index < pixelOrBlockCount; ++index)
-					{
-						const auto posXOfTexelOrBlock = static_cast<uint32_t>(index % size_t(imgBufferWidth / sourceTexelOrBlockStride));
-						const auto posYOfTexelOrBlock = static_cast<uint32_t>(index / size_t(imgBufferWidth / sourceTexelOrBlockStride));
-
-						memcpy
-						(
-							reinterpret_cast<uint8_t*>(layerData) + (posYOfTexelOrBlock * textureGliStrideInBytes),	// copy to a beginning of a certain row
-							sourceData + (posYOfTexelOrBlock * imgBufferWidthInBytes),								// get adjusted beginning of a certain row using to copying process
-							textureGliStrideInBytes																    // get gli row stride with no pitch
-						);
-					}
+					for (size_t zPos = 0; zPos < textureGLIDepth; ++zPos)
+						for (size_t yPos = 0; yPos < imgBufferHeight; ++yPos)
+						{
+							const size_t shiftAdressingRowGLI = ((zPos * textureGliImgHeight + yPos) * textureGliStride) * texelBlockByteSize;
+							const size_t shiftAdressingRowIRR = ((zPos * imgBufferHeight + yPos) * imgBufferWidth) * texelBlockByteSize;
+
+							memcpy
+							(
+								reinterpret_cast<uint8_t*>(layerData) + shiftAdressingRowGLI,	// copy to a beginning of a certain row
+								sourceData + shiftAdressingRowIRR,								// get adjusted beginning of a certain row using to copying process
+								textureGliStrideInBytes									        // get gli row stride with no pitch
```

Comment:
still need to work it out from region's extent (and clamp it appropriately so you don't go out of bounds if sink is too small)